embassy-nrf/src/rng.rs

```diff
@@ -0,0 +1,218 @@
+use core::cell::RefCell;
+use core::convert::Infallible;
+use core::future::Future;
+use core::marker::PhantomData;
+use core::ptr::NonNull;
+use core::task::Poll;
+use core::task::Waker;
+
+use embassy::interrupt::InterruptExt;
+use embassy::traits;
+use embassy::util::CriticalSectionMutex;
+use embassy::util::OnDrop;
+use embassy::util::Unborrow;
+use embassy_extras::unborrow;
+use futures::future::poll_fn;
+use rand_core::RngCore;
+
+use crate::interrupt;
+use crate::pac;
+use crate::peripherals::RNG;
+
+impl RNG {
+    fn regs() -> &'static pac::rng::RegisterBlock {
+        unsafe { &*pac::RNG::ptr() }
+    }
+}
+
+static STATE: CriticalSectionMutex<RefCell<State>> =
```

Comment:
You can make this totally zero-cost.

- for `waker` use `AtomicWaker`
- for `buffer` use two `AtomicPtr<u8>`: `start`, `end`.

This makes it `Sync`, so you can directly put it in a `static` without Mutex/RefCell.

When a byte is ready, write it to `start`, and increment `start`. You're done when `start == end`.

`index` is no longer needed, so as a bonus State is now 3 words instead of 4.